grpc/grpc.go

```diff
@@ -0,0 +1,132 @@
+package grpc
+
+import (
+	"bufio"
+	"encoding/binary"
+	"io"
+	"net"
+	"net/http"
+	"strings"
+)
+
+// based on https://github.com/grpc/grpc-go/blob/master/internal/transport/handler_server.go and https://github.com/Dreamacro/clash/blob/master/transport/gun/gun.go
+
+const (
+	grpcContentType = "application/grpc"
+)
+
+func IsGRPC(w http.ResponseWriter, r *http.Request) bool {
+	if _, ok := w.(http.Flusher); !ok {
+		return false
+	}
+	if r.Method != http.MethodPost {
+		return false
+	}
+	if r.ProtoMajor != 2 {
+		return false
+	}
+	if !strings.HasPrefix(r.Header.Get("Content-Type"), grpcContentType) {
+		return false
+	}
+	return true
+}
+
+type Conn struct {
+	closer        io.Closer
+	bufReadWriter *bufio.ReadWriter
+	remaining     int
+	header        http.Header
+	flusher       http.Flusher
+	headerWritten bool
+
+	closed bool
+}
+
+// GRPC with protobuf is a TLV protocol, 5 bytes header and tlv
+// first byte 0 means uncompressed, 1 means compressed, the next four is the payload length of the total message
+
+// 6th byte is the tag, with v2ray's implementation is value 10
+// next several bytes is unsigned variables length int
+// remaining is the actual payload
+func NewConn(request *http.Request, responseWriter http.ResponseWriter) *Conn {
+	return &Conn{
+		closer:        request.Body,
+		bufReadWriter: bufio.NewReadWriter(bufio.NewReader(request.Body), bufio.NewWriter(responseWriter)),
```

Comment:
Is is necessary to wrap request and response with `bufio.ReadWriter`

Reply:
bufio.Reader is neccessary to read the tlv. bufio.Writer is just for convenient purpose. Clash is using bytes.Buffer to encode tlv header.

Reply:
Not everyone use clash. Clash should not be the reason.

Reply:
I'm quoting clash as it use bytes.Buffer to encode the grpc header. This implentation works with v2ray client as well as I have tested.

Reply:
When creating a bufio.Reader, a 4k size byte slice is created. I do prefer implementing tlv reader from r.Body by myself to avoid creating a slice.